disco/extensions/upgrade_simulation/upgrades/automated_thermal_upgrades.py

```diff
@@ -0,0 +1,280 @@
+
+        xfmr_loading_df = get_all_transformer_info(
+            compute_loading=True, upper_limit=thermal_config["xfmr_upper_limit"]
+        )
+        overloaded_xfmr_list = list(
+            xfmr_loading_df.loc[xfmr_loading_df["status"] == "overloaded"][
+                "name"
+            ].unique()
+        )
+        logger.info("Determined xfmr loadings.")
+        logger.info(f"Number of xfmr violations: {len(overloaded_xfmr_list)}")
+        
+        # xfmr_loading_df = get_all_transformer_info(
+        #     compute_loading=True, upper_limit=thermal_config["xfmr_upper_limit"]
+        # )
+        # overloaded_xfmr_list = list(
+        #     xfmr_loading_df.loc[xfmr_loading_df["status"] == "overloaded"][
+        #         "name"
+        #     ].unique()
+        # )
+        if len(overloaded_xfmr_list) > prev_num_xfmr_violations:
+            logger.info("Write upgrades till this step in debug_upgrades.dss")
+            # self.write_dat_file(output_path=os.path.join(self.config["Outputs"], "debug_upgrades.dss"))
+            raise Exception(
+                f"Xfmr violations increased from {prev_num_xfmr_violations} to {len(overloaded_xfmr_list)} "
+                f"during upgrade process"
+            )
+        if len(overloaded_xfmr_list) > 0:
+            xfmr_commands_list, temp_xfmr_upgrades_df = correct_xfmr_violations(
+                xfmr_loading_df=xfmr_loading_df,
+                xfmr_design_pu=thermal_config["xfmr_design_pu"],
+                xfmr_upgrade_options=xfmr_upgrade_options,
+                parallel_xfmrs_limit=thermal_config["parallel_xfmrs_limit"]
+            )
+
+            logger.info("Corrected xfmr violations.")
+            commands_list = commands_list + xfmr_commands_list
+            xfmr_upgrades_df = xfmr_upgrades_df.append(temp_xfmr_upgrades_df)
+
+        # compute loading after upgrades
+        xfmr_loading_df = get_all_transformer_info(
+            compute_loading=True, upper_limit=thermal_config["xfmr_upper_limit"]
+        )
+        overloaded_xfmr_list = list(
+            xfmr_loading_df.loc[xfmr_loading_df["status"] == "overloaded"][
+                "name"
+            ].unique()
+        )
+        line_loading_df = get_all_line_info(
+            compute_loading=True,
+            upper_limit=thermal_config["line_upper_limit"],
+            ignore_switch=ignore_switch,
+        )
+        overloaded_line_list = list(
+            line_loading_df.loc[line_loading_df["status"] == "overloaded"][
+                "name"
+            ].unique()
+        )
+
+        num_violations_curr_itr = len(overloaded_xfmr_list) + len(overloaded_line_list)
+        logger.info(
+            f"Number of devices with violations after iteration {iteration_counter}: {num_violations_curr_itr}"
+        )
+        iteration_counter += 1
+        if iteration_counter > max_upgrade_iteration:
+            logger.info("Max iterations limit reached, quitting")

```

Comment:
Is this an error? Is there a difference in the result between this case and a normal exit of the loop?